pandas/tests/indexing/multiindex/test_multiindex.py

```diff
@@ -271,3 +271,12 @@ def test_multiindex_assign_aligns_as_implicit_tuple(self):
         df1["C"] = s1
         tm.assert_frame_equal(df1, df2)
         tm.assert_frame_equal(df1, df3)
+
+        # GH 62518
+        df4 = DataFrame(
+            columns=MultiIndex.from_arrays(
+                [["a", "a", "z", "z"], pd.Categorical([1, 2, 1, 2])],
+            ),
+            dtype=object,
+        )
+        df4["z"] = df4["z"].astype("int64")
```

Comment:
Could you make this a separate test that uses the exact code snippet in the original issue?